File: 1325-path-with-maximum-probability/path-with-maximum-probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start, int end) {




        // dijikstras 

        vector<pair<int,double>>adj[n];
        for(int i=0;i<edges.size();i++){
            adj[edges[i][0]].push_back({edges[i][1],succProb[i]});
            adj[edges[i][1]].push_back({edges[i][0],succProb[i]});
        }

        vector<double>dist(n,0);
        priority_queue<pair<double,int>>st;
        st.push({1,start});
        dist[start]=1;
        vector<int> seen(n, 0);


        while(!st.empty()){
            auto t=st.top();
            double prob=t.first;
            int node=t.second;
            st.pop();
            
   if(!seen[node]) {
                seen[node]++;

            for(auto child:adj[node]){
                
                int childnode=child.first;
                double childprob=child.second;
                cout<<childprob<<endl;
                if(childprob*prob>dist[childnode]){
                    dist[childnode]=childprob*prob;
                    st.push({childprob*prob,childnode});
                    // for(int i=0;i<n;i++){
                    //     cout<<dist[i]<<" ";
                    // }cout<<endl;
                }
            }
   }


        }



        return dist[end];
        
    }
};

File: 1325-path-with-maximum-probability/path-with-maximum-probability.py
class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
        graph = defaultdict(list)
        for i, (s, e) in enumerate(edges):
            graph[s].append((e, succProb[i]))
            graph[e].append((s, succProb[i]))
        n = len(graph)
        heap = [(-1, start)]
        seen = set()
        while heap:
            prob, node = heapq.heappop(heap)
            if node == end:
                return -prob
            seen.add(node)
            for nextNode, weight in graph[node]:
                if nextNode in seen:
                    continue
                heapq.heappush(heap, (prob * weight, nextNode))
        return 0

        

        